### PR TITLE
chore(customizer): add missing configs + update validations

### PR DESCRIPTION
**Pull request by anubhutivyas**
## Summary
- Closes the gap between what the Automodel and NeMo-RL frameworks expose in their reference recipes and what the Customizer API accepts, and fixes a LoRA bug that made `exclude_modules` unusable.
- Audited the RL/Automodel recipes for Qwen3 / Qwen3-MoE and Nemotron-3.5-Lightning against our job schemas. 

## Changes
- **Bug fix:** `_translate_lora_config` backfilled `target_modules` whenever it was empty, even when the caller had supplied `exclude_modules`. Automodel's `PeftConfig` rejects both together, so any job naming `exclude_modules` failed *inside the training container* with "target_modules and exclude_modules are mutually exclusive", after it had already claimed GPUs.

- **New Automodel fields** — `training.activation_checkpointing`, `training.mtp.*` (Multi-Token Prediction), `training.backend.*` (per-component kernel selection), `parallelism.pipeline.*` (PP schedule), `training.lora.use_memory_efficient_lora`, `batch.packed_sequence_size`, `dataset.shuffle`.

- **New NeMo-RL fields** — `optimizer_name` + `optimizer_kwargs` (reaches TE `FusedAdam` with bf16 optimizer state, a real memory difference at 30B), `env_vars`, `vllm_kwargs`, `logprob_chunk_size`, `moe_parallelizer`.

## Type of Change

- [ ] Code change (feature, bug fix, or refactor)
- [ ] Code change with documentation updates
- [ ] Documentation only
- [ ] Contributor tooling or automation
- [ ] CI, build, or test infrastructure

## Quality Gates

- [ ] Tests added or updated for changed behavior
- [ ] Existing tests cover changed behavior — justification:
- [ ] Tests not applicable — justification:
- [ ] Documentation updated for user-visible behavior
- [ ] Documentation not applicable — justification:

## Verification

- [ ] Pull request title follows the repository's Conventional Commit format
- [ ] Every commit includes an appropriate `Signed-off-by:` trailer
- [ ] `uv run pre-commit run -a` passes, or any blocked checks are identified below
- [ ] Targeted tests pass, or tests are marked not applicable above
- [ ] No secrets, API keys, or credentials are included

Targeted validation:

## Summary by CodeRabbit

* **New Features**
  * Added options for memory-efficient LoRA, dataset shuffling, activation checkpointing, multi-token prediction, backend optimization, packed sequences, and pipeline parallelism.
  * Added optimizer overrides, worker environment variables, vLLM and MoE settings, and log-probability chunking for DPO and GRPO jobs.
  * Added optional full NeMo-Gym result-table logging for GRPO with W&B integration.

* **Validation**
  * Added checks for incompatible LoRA filters and invalid backend, pipeline, and logging configurations.

* **Documentation**
  * Expanded CLI reference documentation for Automodel cluster job submission options.

### CODE REVIEW

File: services/automodel/src/nmp/automodel/api/v2/jobs/schemas.py [hunk 1]
```diff
@@ -202,13 +216,72 @@ class ParallelismParams(BaseModel):
...
 
 
+class BackendParams(BaseModel):
+    """Which implementation Automodel uses for each model component.
+
+    Fields default to ``None`` meaning "leave it to Automodel", whose own defaults depend
+    on what the training node provides; only explicitly set values are forwarded.
+    """
+
+    attn: Optional[Literal["te", "sdpa", "flex", "eager", "tilelang", "cudnn"]] = Field(
```
Comment (soluwalana) [inline services/automodel/src/nmp/automodel/api/v2/jobs/schemas.py]: Do we want to leave all of these with no default, or would it make more sense to pull the defaults from the code up to the schema so the end user knows what they are getting?

**Review by soluwalana — APPROVED** [review]

File: services/automodel/src/nmp/automodel/api/v2/jobs/schemas.py (hunk 1 above)
Comment (anubhutivyas) [reply services/automodel/src/nmp/automodel/api/v2/jobs/schemas.py]: For the defaults, Automodel picks these configs like `attn`, `linear`, `rope_fusion`, `experts` and `dispatcher` at runtime, based on whether TE, CUDA and DeepEP are present on the training node.
And also there's no single set of defaults that works for all models, so I left then `None`. If user know what they're doing they can set them, else we let automodel pick one.

Comment (soluwalana) [reply services/automodel/src/nmp/automodel/api/v2/jobs/schemas.py]: Yeah, I get the reasoning there, just wish that there was a way to expose the defaults selected. Otherwise the same payload for training can drift substantially inexplicably, which can be frustrating.

Comment (soluwalana) [reply services/automodel/src/nmp/automodel/api/v2/jobs/schemas.py]: I'm fine if you want to create a Linear ticket and we can discuss strategies at the next planning meeting.

Comment (anubhutivyas) [reply services/automodel/src/nmp/automodel/api/v2/jobs/schemas.py]: Agreed, created the ticket here - https://linear.app/nvidia/issue/AALGO-727/add-defaults-for-backendparams-in-automodel-job-config

**Review by ironcommit — APPROVED** [review]

**Review by soluwalana — APPROVED** [review]

File: services/automodel/src/nhx/automodel/api/v2/jobs/schemas.py [hunk 2]
```diff
@@ -273,6 +346,29 @@ class _TrainingBase(BaseModel):
         gt=0,
         description="Samples analyzed to estimate the optimal pack size when sequence packing is enabled.",
     )
+    packed_sequence_size: Optional[int] = Field(
+        default=None,
+        gt=0,
+        description="Pin the packed sequence length instead of estimating it. Requires sequence_packing.",
+    )
```
[the remark this answers is not in the document]
Comment (anubhutivyas) [reply services/automodel/src/nhx/automodel/api/v2/jobs/schemas.py]: fixed

File: services/automodel/src/nhx/automodel/tasks/training/backends/config.py [hunk 3]
```diff
@@ -372,6 +407,14 @@ def _configure_sequence_packing(
     if not customizer_config.batch.sequence_packing:
         return customizer_config.model.max_seq_length, None
 
+    explicit_pack_size = customizer_config.batch.packed_sequence_size
+    if explicit_pack_size is not None:
+        # The caller pinned the length, so skip the sampling pass entirely. packing_factor
+        # stays None, which the schedule reads as "do not rescale the step count".
+        logger.info(f"Sequence packing enabled with caller-supplied pack_size={explicit_pack_size}")
+        cfg["packed_sequence"] = {"packed_sequence_size": explicit_pack_size}
+        return explicit_pack_size, None
```
[the remark this answers is not in the document]
Comment (anubhutivyas) [reply services/automodel/src/nhx/automodel/tasks/training/backends/config.py]: fixed

File: services/rl/src/nhx/rl/tasks/training/backends/nemo_rl/grpo_config.py [hunk 4]
```diff
@@ -113,6 +113,13 @@ def _build_dtensor_cfg(
             # There is no case where recomputing it is what a caller wants, so this is not
             # a knob: NeMo-RL's own MoE recipes set it wherever AC is on.
             dtensor_cfg["moe_parallelizer"] = {"ignore_router_for_ac": True}
+    if grpo_hp.moe_parallelizer:
+        # Merged rather than replaced so a caller reaching for another parallelizer knob
+        # does not silently drop ignore_router_for_ac and hit the recompute crash.
+        dtensor_cfg["moe_parallelizer"] = {
+            **dtensor_cfg.get("moe_parallelizer", {}),
+            **grpo_hp.moe_parallelizer,
```
[the remark this answers is not in the document]
Comment (anubhutivyas) [reply services/rl/src/nhx/rl/tasks/training/backends/nemo_rl/grpo_config.py]: fixed

File: services/rl/src/nhx/rl/schemas/job.py [hunk 5]
```diff
@@ -770,6 +816,16 @@ def validate_for_training(self) -> None:
         if isinstance(training, GRPOTraining):
             if not self.environment:
                 raise ValueError("GRPO jobs require an environment fileset reference.")
+            # NeMo-RL gates the Tables on `wandb_enabled AND the flag`, so without the
+            # integration this is a silent no-op.
+            if training.log_nemo_gym_full_result_tables and (
+                self.integrations is None or self.integrations.wandb is None
+            ):
```
[the remark this answers is not in the document]
Comment (anubhutivyas) [reply services/rl/src/nhx/rl/schemas/job.py]: fixed

File: services/rl/src/nhx/rl/schemas/job.py [hunk 6]
```diff
@@ -505,6 +524,23 @@ class GRPOTraining(_TrainingBase):
...
+        default=None,
+        description="Additional arguments for the vLLM rollout engine, for example "
+        "{'moe_backend': 'triton'}. Forwarded to vLLM as supplied.",
+    )
+    moe_parallelizer: dict[str, Any] | None = Field(
+        default=None,
+        description="Merged over policy.dtensor_cfg.moe_parallelizer, which controls how MoE experts "
+        "are sharded and checkpointed. The platform already sets ignore_router_for_ac when expert "
+        "parallelism and activation checkpointing are both on; this reaches the rest.",
+    )
```
[the remark this answers is not in the document]
Comment (anubhutivyas) [reply services/rl/src/nhx/rl/schemas/job.py]: fixed

File: services/rl/src/nhx/rl/schemas/job.py [hunk 7]
```diff
@@ -185,6 +185,20 @@ class _TrainingBase(RlSchema):
         description="Optimizer + LR-scheduler combination (AdamW/Adam × cosine-annealing/flat-LR). "
         "Defaults to AdamW with cosine annealing.",
     )
+    optimizer_name: str | None = Field(
+        default=None,
+        description="Import path of the optimizer class, overriding the one implied by optimizer_type "
+        "(e.g. 'transformer_engine.pytorch.optimizers.fused_adam.FusedAdam'). optimizer_type still "
+        "selects the LR schedule. Unset uses torch.optim.AdamW / Adam.",
+    )
```
[the remark this answers is not in the document]
Comment (anubhutivyas) [reply services/rl/src/nhx/rl/schemas/job.py]: fixed
